SCClassLibrary/Common/Core/CondVar.sc

```diff
@@ -0,0 +1,126 @@
+CondVar {
+	var waitingThreads;
+
+	*new { ^super.newCopyArgs(Array.new, Array.new) }
+
+	wait { |predicate|
+		if(predicate.isNil) {
+			this.prWait
+		} {
+			// If the predicate is already true, we return immediately. If not,
+			// we block until the predicate is true after being signalled.
+			while { predicate.value.not } { this.prWait }
+		}
+	}
+
+	waitFor { |timeoutBeats, predicate|
+		var endingBeats;
+
+		// Convert timeoutBeats to a value we know is safe to call `wait` on,
+		// since an exception thrown in the timeout thread is a very bad thing.
+		// May throw if preconditions are not met.
+		timeoutBeats = this.prHandleTimeoutPreconditions(timeoutBeats);
+
+		// This is a slightly more complex version of the body of `wait`. The
+		// extra complexity comes from maintaining the timeout requirement.
+		if(predicate.isNil) {
+			// Interface requirement: immediately return if timeout is non-positive
+			if(timeoutBeats <= 0) { ^false };
+			^this.prWaitFor(timeoutBeats)
+		} {
+			// By taking thisThread.beats here, we wait a little longer than we
+			// should because of the time it takes to get from the start of `waitFor`
+			// to this line, but that's alright since precise timeouts aren't an
```

Comment:
Just a spot comment for now -- `thisThread.beats` is logical time, not physical time. Between requesting the wait/timeout and calculating the ending beads, there is no `yield`, so logical time advances by 0. So it won't "wait a little longer" -- measured in terms of logical time, it should be exact. (Measured by physical time, there would be some deviation, but that's true of a simple `dur.wait` as well.)

Reply:
thanks for pointing this out! I forgot SC's clocks have this concept. I've updated my code to reflect this. I thought through the implementation again with this new information and I don't think it requires any code changes. I'd appreciate if you could suggest if/how I should note this in the documentation.

this now opens up the possibility that a predicate could modify thisThread.beats and muck with the waitFor timeout logic. I don't know if this is a case worth worrying about, I'd also be happy to just stick a note in the documentation to the effect of "do this at your own risk". what do you think?

Reply:
I agree -- if anyone starts messing around with the thread's or the clock's beats, there is no longer any guarantee of time values advancing in a natural way. (I'd take this as a natural enough assumption that it might not even need to be documented.)

Hoping to get a chance to look at this more carefully over the weekend.